dc/s2n-quic-dc/events/acceptor.rs

```diff
@@ -0,0 +1,343 @@
+    ///
+    /// If streams consume more time than this value to initialize, they
+    /// may potentially be replaced by more recent streams.
+    #[measure("max_sojourn_time", Duration)]
+    max_sojourn_time: core::time::Duration,
+}
+
+/// Emitted when a fresh TCP stream is enqueued for processing
+#[event("acceptor:tcp:fresh:enqueued")]
+#[subject(endpoint)]
+struct AcceptorTcpFreshEnqueued<'a> {
+    /// The remote address of the TCP stream
+    #[builder(&'a s2n_quic_core::inet::SocketAddress)]
+    remote_address: SocketAddress<'a>,
+}
+
+/// Emitted when a the TCP acceptor has completed a batch of stream enqueues
+#[event("acceptor:tcp:fresh:batch_completed")]
+#[subject(endpoint)]
+struct AcceptorTcpFreshBatchCompleted {
+    /// The number of fresh TCP streams enqueued in this batch
+    #[measure("enqueued")]
+    enqueued: usize,
+
+    /// The number of fresh TCP streams dropped in this batch due to capacity limits
+    #[measure("dropped")]
+    dropped: usize,
+
+    /// The number of TCP streams that errored in this batch
+    #[measure("errored")]
+    errored: usize,
+}
+
+/// Emitted when a TCP stream has been dropped
+#[event("acceptor:tcp:stream_dropped")]
+#[subject(endpoint)]
+struct AcceptorTcpStreamDropped<'a> {
+    /// The remote address of the TCP stream
+    #[builder(&'a s2n_quic_core::inet::SocketAddress)]
+    remote_address: SocketAddress<'a>,
+
+    #[nominal_counter("reason")]
+    reason: AcceptorTcpStreamDropReason,
+}
+
+enum AcceptorTcpStreamDropReason {
+    /// There were more streams in the TCP backlog than the userspace queue can store
+    FreshQueueAtCapacity,
+
+    /// There are no available slots for processing
+    SlotsAtCapacity,
+}
+
+/// Emitted when a TCP stream has been replaced by another stream
+#[event("acceptor:tcp:stream_replaced")]
+#[subject(endpoint)]
+struct AcceptorTcpStreamReplaced<'a> {
+    /// The remote address of the stream being replaced
+    #[builder(&'a s2n_quic_core::inet::SocketAddress)]
+    remote_address: SocketAddress<'a>,
+
+    /// The amount of time that the stream spent in the accept queue before
+    /// being replaced with another
+    #[timer("sojourn_time")]
+    sojourn_time: core::time::Duration,
+
+    /// The amount of bytes buffered on the stream
+    #[measure("buffer_len", Bytes)]
+    buffer_len: usize,
+}
+
+/// Emitted when a full packet has been received on the TCP stream
+#[event("acceptor:tcp:packet_received")]
+#[subject(endpoint)]
+struct AcceptorTcpPacketReceived<'a> {
+    /// The address of the packet's sender
+    #[builder(&'a s2n_quic_core::inet::SocketAddress)]
+    remote_address: SocketAddress<'a>,
+
+    /// The credential ID of the packet
+    #[snapshot("[HIDDEN]")]
+    credential_id: &'a [u8],
+
+    /// The stream ID of the packet
+    stream_id: u64,
+
+    /// The payload length of the packet
+    #[measure("payload_len", Bytes)]
+    payload_len: usize,
+
+    /// If the packet includes the final bytes of the stream
```

Comment:
Is that right? Do we want to distinguish between final bytes arrived vs. length known upfront?

Reply:
Good question. Currently the former is what's being computed. But I can add another field to distinguish the packet carrying and final offset.

Reply:
Added in https://github.com/aws/s2n-quic/pull/2379/commits/43e8cb7bf1fb353ee40a7b113e6f38296bcf8f30#diff-9d58bbb85d84818da3d635f23f8c8b06cff3ac114f8a89f51d130250226a7c32R136